airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -324,19 +346,42 @@ def read_records(
             stream_slice=_slice,
             records_schema=records_schema,
         )
-        for stream_data in self._read_pages(record_generator, self.state, _slice):
-            current_record = self._extract_record(stream_data, _slice)
-            if self.cursor and current_record:
-                self.cursor.observe(_slice, current_record)
 
-            # Latest record read, not necessarily within slice boundaries.
-            # TODO Remove once all custom components implement `observe` method.
-            # https://github.com/airbytehq/airbyte-internal-issues/issues/6955
-            most_recent_record_from_slice = self._get_most_recent_record(most_recent_record_from_slice, current_record, _slice)
-            yield stream_data
+        if self.cursor and isinstance(self.cursor, ResumableFullRefreshCursor):
+            stream_state = self.state
 
-        if self.cursor:
-            self.cursor.close_slice(_slice, most_recent_record_from_slice)
+            # Before syncing the RFR stream, we check if the job's prior attempt was successful and don't need to fetch more records
+            # The platform deletes stream state for full refresh streams before starting a new job, so we don't need to worry about
+            # this value existing for the initial attempt
+            if stream_state.get(FULL_REFRESH_SYNC_COMPLETE_KEY):
+                return
+            cursor_value = stream_state.get("next_page_token")
+
+            # The first attempt to read a page for the current partition should have the paginator reset to the current
+            # cursor state which is initially set to the incoming state from the platform
+            partition_key = self._to_partition_key(_slice.partition)
+            if partition_key not in self._synced_partitions:
+                self._synced_partitions[partition_key] = True
```

Comment:
Hmmm let me try to articulate it in an example, I think you have it mostly right. For our purposes yes lets just stick with one partition, although it is designed to work for parent stream records.

On the first attempt of the job, `_synced_partitions` is empty. So on the first iteration we pass the check since we know we haven't synced this partition yet. And we reset the value back to 0. It was already that, but we just reset it to itself. Then we continue paginating and we never reset because the partition is in the map. Now let's say we fail on page 10 and we've written state up to this point to page 9.

On the second attempt of the job `_synced_partitions` is empty so we reset this to the previous state of page 9. We then start paginating which has now been reset to page 9 until we complete.

On the next full refresh job, no state comes in so we paginate beginning at 0.

Does that help clarify things or did it just make it more confusing?

Reply:
>  Then we continue paginating and we never reset because the partition is in the map.

I think this is what was confusing me, because `synced` to me indicates that we're _done_ syncing it, as in something that would be set at the end instead of the beginning. What about `started_partitions` since the act of resetting to the initial state is associated with starting to sync a partition? Or to Alex's point about it not being obvious, maybe `_rfr_initialized_partitions`?

> On the second attempt of the job _synced_partitions is empty so we reset this to the previous state of page 9. We then start paginating which has now been reset to page 9 until we complete.

This is definitely helpful on the logic side! Just the naming is caught for me now but I understand how its implemented.